source/extensions/clusters/dynamic_forward_proxy/cluster.h

```diff
@@ -145,13 +173,29 @@ class Cluster : public Upstream::BaseDynamicClusterImpl,
   const envoy::config::endpoint::v3::LocalityLbEndpoints dummy_locality_lb_endpoint_;
   const envoy::config::endpoint::v3::LbEndpoint dummy_lb_endpoint_;
   const LocalInfo::LocalInfo& local_info_;
+  Event::Dispatcher& main_thread_dispatcher_;
+  const envoy::config::cluster::v3::Cluster orig_cluster_config_;
+
+  Event::TimerPtr idle_timer_;
 
   // True if H2 and H3 connections may be reused across different origins.
   const bool allow_coalesced_connections_;
 
   mutable absl::Mutex host_map_lock_;
   HostInfoMap host_map_ ABSL_GUARDED_BY(host_map_lock_);
 
+  mutable absl::Mutex cluster_map_lock_;
+  ClusterInfoMap cluster_map_ ABSL_GUARDED_BY(cluster_map_lock_);
+
+  // only for using tls_.isShutdown() in Cluster::~Cluster().
+  ThreadLocal::TypedSlot<ThreadLocalConfig> tls_;
```

Comment:
This is kind of strange if only used for the shutdown case. Since you are passed the cluster factory context, can we just expose a method on the server about whether it's shutdown or not (there might already be one) or somehow make the cluster manager just ignore whatever is happening during shutdown to avoid having that leak into this code? Thank you.